lib/activerecord/delay_touching/state.rb

```diff
@@ -33,8 +33,13 @@ def records_by_attrs_and_class
         @records.map { |attrs, records| [attrs, records.group_by(&:class)] }
       end
 
+      # There are more records as long as there is at least one record that is persisted
       def more_records?
-        @records.present?
+        @records.each do |_, set|
```

Comment:
at most, a single iteration over all tracked records to find a persisted record, but shortcut on first identified.